server/balancer.go

```diff
@@ -370,3 +372,262 @@ func (r *replicaChecker) checkBestReplacement(region *RegionInfo) Operator {
 	}
 	return newTransferPeer(region, oldPeer, newPeer)
 }
+
+// RegionStat records each hot region's statistics
+type RegionStat struct {
+	RegionID       uint64    `json:"region_id"`
+	WrittenBytes   uint64    `json:"written_bytes"`
+	HotDegree      int       `json:"hot_degree"`
+	LastUpdateTime time.Time `json:"last_update_time"`
+	StoreID        uint64    `json:"-"`
+	antiCount      int
+	version        uint64
+}
+
+// ListRegionsStat is a list of a group region state type
+type ListRegionsStat []RegionStat
+
+func (m ListRegionsStat) Len() int           { return len(m) }
+func (m ListRegionsStat) Swap(i, j int)      { m[i], m[j] = m[j], m[i] }
+func (m ListRegionsStat) Less(i, j int) bool { return m[i].WrittenBytes < m[j].WrittenBytes }
+
+// StoreHotRegions records all hot regions in one store with sequence
+type StoreHotRegions struct {
+	TotalWrittenBytes uint64          `json:"total_written"`
+	RegionCount       int             `json:"region_count"`
+	RegionsStat       ListRegionsStat `json:"status"`
+}
+
+type balanceHotRegionScheduler struct {
+	opt         *scheduleOption
+	limit       uint64
+	scoreStatus map[uint64]*StoreHotRegions // store id -> regions status in this store
+	r           *rand.Rand
+}
+
+func newBalanceHotRegionScheduler(opt *scheduleOption) *balanceHotRegionScheduler {
+	return &balanceHotRegionScheduler{
+		opt:         opt,
+		limit:       1,
+		scoreStatus: make(map[uint64]*StoreHotRegions),
+		r:           rand.New(rand.NewSource(time.Now().UnixNano())),
+	}
+}
+
+func (l *balanceHotRegionScheduler) GetName() string {
+	return "balance-hot-region-scheduler"
+}
+
+func (l *balanceHotRegionScheduler) GetResourceKind() ResourceKind {
+	return priorityKind
+}
+
+func (l *balanceHotRegionScheduler) GetResourceLimit() uint64 {
+	return l.limit
+}
+
+func (l *balanceHotRegionScheduler) Prepare(cluster *clusterInfo) error { return nil }
+
+func (l *balanceHotRegionScheduler) Cleanup(cluster *clusterInfo) {}
+
+func (l *balanceHotRegionScheduler) Schedule(cluster *clusterInfo) Operator {
+	l.calculateScore(cluster)
+	region := l.SelectSourceRegion(cluster)
+	if region == nil {
+		return nil
+	}
+	newLeader := l.selectTransferLeader(region)
+	if newLeader != nil {
+		return newPriorityTransferLeader(region, newLeader)
+	}
+	peer := l.selectTransferPeer(region, cluster)
+	if peer != nil {
+		return newPriorityTransferPeer(region, region.Leader, peer)
+	}
+	return nil
+}
+
+func (l *balanceHotRegionScheduler) clearScore() {
+	l.scoreStatus = make(map[uint64]*StoreHotRegions)
+}
```

Comment:
I don't think we need to abstract a function here. It's only 1 line and be called only at 1 place.